cms/server/contest/handlers/api.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
+
+# Contest Management System - http://cms-dev.github.io/
+# Copyright © 2025 Luca Versari <veluca93@gmail.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""API handlers for CMS.
+
+"""
+
+import ipaddress
+import logging
+
+try:
+    import tornado4.web as tornado_web
+except ImportError:
+    import tornado.web as tornado_web
+
+from cms.db.submission import Submission
+from cms.server import multi_contest
+from cms.server.contest.authentication import validate_login
+from cms.server.contest.submission import \
+    UnacceptableSubmission, accept_submission
+from .contest import ContestHandler
+from ..phase_management import actual_phase_required
+
+logger = logging.getLogger(__name__)
+
+
+class ApiLoginHandler(ContestHandler):
+    """Login handler.
+
+    """
+    @multi_contest
+    def post(self):
+        username = self.get_argument("username", "")
+        password = self.get_argument("password", "")
+
+        try:
+            ip_address = ipaddress.ip_address(self.request.remote_ip)
+        except ValueError:
+            logger.warning("Invalid IP address provided by Tornado: %s",
+                           self.request.remote_ip)
+            return None
+
+        participation, login_data = validate_login(
+            self.sql_session, self.contest, self.timestamp, username, password,
+            ip_address)
+
+        if participation is None:
+            self.json({"error": "Login failed"}, 403)
+        elif login_data is not None:
+            cookie_name = self.contest.name + "_login"
+            self.json({"login_data": self.create_signed_value(
+                cookie_name, login_data).decode()})
+        else:
+            self.json({})
+
+    def check_xsrf_cookie(self):
+        pass
+
+
+class ApiTaskListHandler(ContestHandler):
+    """Handler to list all tasks and their statements.
+
+    """
+    @tornado_web.authenticated
+    @actual_phase_required(0, 3)
+    @multi_contest
+    def get(self):
+        contest = self.contest
+        tasks = dict()
+        for task in contest.tasks:
```

Comment:
The tasks should be presented in the same order as in the web UI. Let us sort by `task.num` here.

Reply:
Ah, I didn't expect sorted relationships :) Good.